### ISSUE TITLE
JSON Output Archive with uint64 serializing as numeric, not string GCC vs MSVC

### ISSUE THREAD

**Issue by xXchillaXx** [USCiLab/cereal#556]
Hello everyone!

Thanks for making and maintaining a great library.

I'm using it to serialize some objects between QT applications running cross-platform (Win/Linux x64) and Python.  

Depending on where I am building, I am getting different results:

Sample code for a small QT application compiled and debugged from both Linux (gcc7) and Windows (MCVC2015 x64)

NOTE: I cross posted here: https://groups.google.com/forum/#!topic/cerealcpp/S1B8k5KOcy8 but it looks like the github issues page has more activity.  Sorry for any confusion

```c++
#include <QCoreApplication>
#include <cereal/archives/json.hpp>
#include <cereal/cereal.hpp>

struct SampleStruct
{
    QString QString_Field;
    quint64 quint64_Field;
    bool bool_Field;
    
    template<class Archive>
    void serialize(Archive& archive)
    {
        archive(
                    CEREAL_NVP(quint64_Field),
                    CEREAL_NVP(bool_Field),
                    CEREAL_NVP(QString_Field)
                    );
    }
};

// ** QString **
template <class Archive>
std::string save_minimal(Archive& ar, const QString& m)
{
    return m.toStdString();
}

template <class Archive>
void load_minimal(Archive& ar, QString& m, const std::string& value)
{
    m = QString(value.c_str());
}

int main(int argc, char *argv[])
{
    QCoreApplication a(argc, argv);
    
    QByteArray payload;
    {
        SampleStruct outputSS;
        outputSS.bool_Field = true;
        outputSS.quint64_Field = 123456;
        outputSS.QString_Field = QString("test test");
        
        std::ostringstream os;
        {
            cereal::JSONOutputArchive ar(os);
            ar( outputSS );
        }
        auto serializedData = os.str();
        std::vector<char> vec(serializedData.begin(), serializedData.end());
        auto size = serializedData.size();
        payload = QByteArray(vec.data(), size);
    }

    
    {
        SampleStruct inputSS;
        std::vector<char> vec(payload.begin(), payload.end());
        std::istringstream is(std::string(vec.begin(), vec.end()));
        {
            cereal::JSONInputArchive ar(is);
            ar( inputSS );
        }
    }
    return a.exec();
}
```

When I run from within Linux, my output serialized string is correctly:

`"{\n    "value0": {\n        "quint64_Field": "123456",\n        "bool_Field": true,\n        "QString_Fi"... (125)`

However, when I run this in Windows, I get this:

`"{\n    "value0": {\n        "quint64_Field": 123456,\n        "bool_Field": true,\n        "QString_Fiel"... (123)`

I see that in the Cereal headers there are many flags for `#ifndef _MSC_VER` but is there anything else I can do to help get the consistent behavior I am looking for?

**Comment by JuliusRapp** [comment]
I can reproduce the behavior that `quint64` is serialized as a string on Linux using Qt 4. This is because `quint64` is an alias for `unsigned long long`. The JSON archive handles `unsigned long long` in [line 304](https://github.com/USCiLab/cereal/blob/develop/include/cereal/archives/json.hpp#L304) (develop branch). The implementation there is based on `std::stringstream` and string output.

I can not test on Windows, but I see two possible explanations for your observation (non-string serialization of `quint64`):

1. `quint64` on Windows may be an alias for `unsigned long` or
2. the `unsigned long` overload in [line 289](https://github.com/USCiLab/cereal/blob/develop/include/cereal/archives/json.hpp#L287), which is protected by `#ifdef _MSC_VER`, may be chosen for `unsigned long long` (including some "conversion").

Below you find a modified example that lets you check the exact type of `quint64` on Windows. If you are lucky then your problem can be solved by choosing, say, `std::uint64_t` over `quint64`. Alternatively, the example below may serve as the basis for a minimal example without the Qt dependency.

```
#include <cstdint>

#include <iostream>
#include <type_traits>

#include <QtCore/QtGlobal>

#include <cereal/archives/json.hpp>
#include <cereal/cereal.hpp>

struct SampleStruct {
  quint64 quint64_field = 1u;
  unsigned long long unsigned_long_long_field = 1u;

  std::uint64_t std_uint64_field = 1u;
  unsigned long unsigned_long_field = 1u;
  unsigned unsigned_field = 1u;

  template<class Archive>
  void serialize(Archive& archive) {
    archive(
      CEREAL_NVP(quint64_field),
      CEREAL_NVP(unsigned_long_long_field),

      CEREAL_NVP(std_uint64_field),
      CEREAL_NVP(unsigned_long_field),
      CEREAL_NVP(unsigned_field)
    );
  }
};


int main() {
  std::cout
    << sizeof(unsigned long long)
    << " == sizeof(unsigned long long)\n"
    << sizeof(unsigned long)
    << " == sizeof(unsigned long)\n"
    << sizeof(unsigned)
    << " == sizeof(unsigned)\n"
    << std::endl;

  std::cout
    << std::is_same<quint64, unsigned long long>{}
    << " == std::is_same<quint64, unsigned long long>{}\n"
    << std::is_same<quint64, unsigned long>{}
    << " == std::is_same<quint64, unsigned long>{}\n"
    << std::is_same<quint64, unsigned>{}
    << " == std::is_same<quint64, unsigned>{}\n"
    << std::endl;

  std::cout
    << std::is_same<std::uint64_t, unsigned long long>{}
    << " == std::is_same<std::uint64_t, unsigned long long>{}\n"
    << std::is_same<std::uint64_t, unsigned long>{}
    << " == std::is_same<std::uint64_t, unsigned long>{}\n"
    << std::is_same<std::uint64_t, unsigned>{}
    << " == std::is_same<std::uint64_t, unsigned>{}\n"
    << std::endl;

  std::cout
    << std::is_same<quint64, std::uint64_t>{}
    << " == std::is_same<quint64, std::uint64_t>{}\n"
    << std::endl;

  SampleStruct foo{};
  {
    cereal::JSONOutputArchive archive(std::cout);
    archive(foo);
  }
}
```

**Comment by xXchillaXx** [comment]
Hi JuliusRapp,

Thank you for your quick reply.  Running your sample in Windows compiled with MSVC2015, I get this output:

```
8 == sizeof(unsigned long long)
4 == sizeof(unsigned long)
4 == sizeof(unsigned)

1 == std::is_same<quint64, unsigned long long>{}
0 == std::is_same<quint64, unsigned long>{}
0 == std::is_same<quint64, unsigned>{}

1 == std::is_same<std::uint64_t, unsigned long long>{}
0 == std::is_same<std::uint64_t, unsigned long>{}
0 == std::is_same<std::uint64_t, unsigned>{}

1 == std::is_same<quint64, std::uint64_t>{}

{
    "value0": {
        "quint64_field": 1,
        "unsigned_long_long_field": 1,
        "std_uint64_field": 1,
        "unsigned_long_field": 1,
        "unsigned_field": 1
    }
}
```

What can we do from here?

**Comment by JuliusRapp** [comment]
Interesting output. So on your system, `unsigned` and `unsigned long` are of the same size (32 bit). As a result, both `std::uint64_t` and `quint64` are an alias of `unsigned long long`, which is 64 bit.

Can you change the member type of your struct from `quint64` to `std::uint64_t`? That should give you non-stringified output on both Windows and Linux.

**Comment by xXchillaXx** [comment]
I can try and do that, however, I believe the stringification of 64-bit values is the correct behavior because (https://tools.ietf.org/html/rfc7159#section-6):

>    Note that when such software is used, numbers that are integers and
>    are in the range [-(2**53)+1, (2**53)-1] are interoperable in the
>    sense that implementations will agree exactly on their numeric
>    values.

Cereal even goes out of its way to  say that it does support this (http://uscilab.github.io/cereal/serialization_archives.html)

> Serializing numeric types with JSON: All numeric types are serialized as numbers except for large types such as long long, unsigned long long, and long double, which are serialized as strings. In JSON, numbers will not have surrounding quotes, while the string representations of the larger types will be surrounded with double quotes.

I believe the linux implementation in Cereal is the correct one and it's only a problem with the MSVC 'fixes'

**Comment by xXchillaXx** [comment]
Running your sample in Linux results in: 
```
8 == sizeof(unsigned long long)
8 == sizeof(unsigned long)
4 == sizeof(unsigned)

1 == std::is_same<quint64, unsigned long long>{}
0 == std::is_same<quint64, unsigned long>{}
0 == std::is_same<quint64, unsigned>{}

0 == std::is_same<std::uint64_t, unsigned long long>{}
1 == std::is_same<std::uint64_t, unsigned long>{}
0 == std::is_same<std::uint64_t, unsigned>{}

0 == std::is_same<quint64, std::uint64_t>{}

{
    "value0": {
        "quint64_field": "1",
        "unsigned_long_long_field": "1",
        "std_uint64_field": 1,
        "unsigned_long_field": 1,
        "unsigned_field": 1
    }
}
```

**Comment by JuliusRapp** [comment]
> [...] I believe the stringification of 64-bit values is the correct [...]

Oh, now I understand that you have your reasons to prefer the stringified version. Good catch!



> Cereal even goes out of its way to say that it does support this (http://uscilab.github.io/cereal/serialization_archives.html)
> 
> > Serializing numeric types with JSON: All numeric types are serialized as numbers except for large types such as long long, unsigned long long, and long double, which are serialized as strings. In JSON, numbers will not have surrounding quotes, while the string representations of the larger types will be surrounded with double quotes.

The problem in this quote is the unknown `sizeof(unsigned long long)`. Unsigned 64-bit integers are purely forwarded by JSONOutputArchive to the RapidJSON writer in [line 252](https://github.com/USCiLab/cereal/blob/develop/include/cereal/archives/json.hpp#L252). As far as I see, RapidJSON does not perform any string conversions either, see [here](https://github.com/USCiLab/cereal/blob/develop/include/cereal/external/rapidjson/writer.h#L177) and [here](https://github.com/USCiLab/cereal/blob/develop/include/cereal/external/rapidjson/writer.h#L299).



> I believe the linux implementation in Cereal is the correct one and it's only a problem with the MSVC 'fixes'

After reading the JSON specification I agree that there seems to be a problem with non-stringified 64-bit integers in JSON. However, I think that the GCC handling of `unsigned long long` is just "correct" by accident.



The bottom line is that there might be a problem in RapidJSON because it serializes 64-bit integers as plain numbers, but there is no guarantee that this works beyond 53 bits. In addition, one could think about the correct handling of `unsigned long long` if it has the same size as `unsigned long` (Linux).

I can not help with any of these problems, sorry. Anyone else?

**Comment by xXchillaXx** [comment]
are there any macros or helpers I can use within my `serialize` functions to force the type to be stringified?

**Comment by xXchillaXx** [comment]
> I can not help with any of these problems, sorry. Anyone else?

Thank you so much for your help with my issue, regardless :+1: 

I think based on these assumptions, we will try to either use the behavior and ensure that the values transmitted are less than 2^53 or we will explicitly pass a field with **_str** suffix and convert to/from 64-bit integer as needed (it will at least be more explicit in the python side)
